New ui provide info or warning messages to users, related to A095/A098/A096/S055

juzen2003 opened the pull request:
# Update:
- A095:
  - If user's browser is older than the following version: Firefox (66+), Chrome (74+), Safari (12.1+), Opera (58+), and Microsoft Edge (18+), we display a modal to inform the user to update the browser version. The modal is not closable, so the user cannot proceed to use OPUS until he/she updates the browser version.
- A098:
  - If browser window width is less than 1280px, we display a modal to inform the user to resize it. The modal will not go away and user cannot click anywhere in OPUS until the browser width is resized to some values larger than 1280px.
- A096:
  - If the widget cookies is undefined, we know the user is a first time visitor. We display a welcome modal for now, but will update to display the real guide page in the future.

juzen2003 committed:
- In base.html, add another confirm modal (#op-browser-warning) to display
  browser warning message when user browser is not supported. (line 345)
- In confirm_modal.html, add if block to not display footer if there is
  no yes/no buttons. Also make it flexible to enable/disable close button
  by passing in allow_close.
- In opus.js, add checkBrowserVersion function to check user's browser, if
  it's not supported, we will display a modal (without yes/no and close buttons)
  to inform user to update browse version.

juzen2003 committed:
- In opus.js, add checkBrowserSize function to check the browser width,
  if it's less than 1280px, we display a modal to inform the user to resize
  the browser.
- In opus.js, update the comments about the functionalities of
  checkBrowserSize and checkBrowserVersion.
- In base.html, add a modal with id "op-browser-size-msg" as the information
  dialog box when the user has a narrow window size (less than 1280px).
- Rename "op-browser-warning" to "op-browser-version-msg".

juzen2003 committed:
- In opus.js, update logic in checkBrowserVersion, replace switch block
  with one if else block by using global variable opus.browserSupport
  (line 785-791).

juzen2003 committed:
- In confirm_modal.html, set data-backdrop to "static". This will enable a
  modal-backdrop element in the background when a confirm_modal is open,
  but will not close the modal when clicking the background.

juzen2003 committed:
- In base.html, add one confirm_modal with id "op-guide" as a temporarily
  display for the first time visitors to OPUS, we will replace this part
  by calling __help/splash.html api in the future.
- In opus.js, add checkCookies function to check if widgets cookies have
  already existed to determine if the user is a first time visitor. If
  so, we display the op-guide modal.

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -724,11 +734,91 @@ var opus = {
+        }
+    },
+
+    checkCookies: function() {
```
We're going to get rid of the widgets cookie, which doesn't currently do anything anyway. We should use some unique cookie for determining if the user is new.

rfrenchseti replied:
That means we should also set the cookie here. Don't rely on the widget code to do it for you.

juzen2003 replied:
Now we check and set visited cookies here.

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -724,11 +734,91 @@ var opus = {
 
 $(document).ready(function() {
+    opus.checkBrowserVersion();
+    opus.checkCookies();
```
This is probably the place you need to above if checkBrowserVersion returns saying the browser is bad. Don't do anything else, including trying to initialize OPUS. Just give up.

rfrenchseti reviewed (changes requested):
- When you decrease the browser size below 1280px, and then keep decreasing it, there's a limit as to how far the dialog will move to the left. It doesn't just stay centered on the screen. Eventually you have a big space on the left and the dialog is cut off on the right. This probably won't work well on a mobile device.
- If you are both a new user and are running an old browser, you get both modal dialogs. But the new user one is on top, so you don't see the browser error. Also when you click OK on that one, both dialogs get closed so you *never* see the one about the browser version. OPUS just fails to work.
- Is there a way to tell the user which browser and version they are currently using?
- Add a catch for the browser being too short of 275px. Change the About message accordingly.

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -336,6 +340,29 @@ <h3 class='op-header-text'></h3>
+        {% include "ui/confirm_modal.html" with
+            id_name="op-browser-size-msg"
+            confirm_title="Message About Your Browser Size"
+            confirm_msg="Please resize your browser. OPUS works best with a browser width of at least 1280px."
```
Also let's spell out the word "pixels" here. Nobody knows what a "px" is. Although pixels is not really technically accurate, at least people will have heard of it.

juzen2003 committed:
- Move constant galleryAndTablePSLength from browse.js to opus.js, and
  it's used in both browse.js and cart.js to set the fixed ps length for
  table and gallery view.

juzen2003 committed:
- In about.html, modify part of the messages to be "OPUS works best with
  a browser size of at least 1280 by 275." Originally, it was 1280 by 720.
- In base.html, modify the confirm message for op-browser-size-msg modal
  to be "OPUS works best with a browser size of at least 1280 pixels by
  275 pixels."
- In opus.js, update checkBrowserSize to display the modal as well when
  browser height is less than 275px. Also call the function at document
  ready such that when the user starts with short browser screen, he/she
  can see the modal as well.

juzen2003 committed:
- In confirm_modal, add "modal-dialog-centered" to display modal in the
  center of the screen.
- In browse.js, disable draggable for the following modals:
  #op-browser-version-msg, #op-browser-size-msg, #op-guide. This is to
  make sure that modal can stay in the center of the screen when the
  browser is narrow.
- In opus.js, update the logic in document ready. If browser version is
  not supported, we display the modal and will not proceed to OPUS
  initialization.

juzen2003 committed:
- In base.html, remove confirm_msg since we will customize it in
  isBrowserSupported function in opus.js.
- In opus.js, rename checkBrowserVersion to isBrowserSupported, add the
  browser name and version as the modal message for #op-browser-version-msg.

juzen2003 committed:
- In opus.js line 687, update the event handler for modal button click.
  Now clicking yes (submit) button in a modal will only close the modal
  itself instead of all the confirm modals.
- In opus.js line 659, update the event handler for esc keydown. Esc
  keydown will not close #op-browser-version-msg and #op-browser-size-msg
  now.

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -724,12 +741,108 @@ var opus = {
+         */
+        if ($.cookie("visited") === undefined) {
+            // set the cookie for the first time user
+            $.cookie("visited", true, { expires: 28});  // days
```
So the way this is currently implemented, a user can use OPUS every day  and then on the 28th day, get the splash dialog again, right? That seems wrong. We should either make the cookie never expire, or we should reset the 28 day counter every time the user runs OPUS (set it outside of the if).

rfrenchseti commented on opus/application/apps/help/templates/help/about.html:
```diff
@@ -176,7 +176,7 @@
 </p>
 
-<p>Please note that OPUS works best with a browser size of at least 1280 by 720. We
+<p>Please note that OPUS works best with a browser size of at least 1280 by 275. We
```
275 is not working the way I expected. Perhaps I screwed up using the inspector tool. The idea is that in the gallery view with the metadata detail modal popped up, it should fit completely on the screen. Otherwise, if the browser is too short, a bogus scrollbar shows up. So figure out how tall that is, round it to something pretty, and use that value.

juzen2003 committed:
- In about.html, update about message to "OPUS works best with a browser
  size of at least 1280 by 720".
- In opus.js, update checkBrowserSize, it will now check for height 720
  pixels, if screen height is less than 720 pixels, a modal will show up.
- In base.html, remove confirm_msg in op-browser-size-msg modal, modal
  message will now be updated in checkBrowserSize.

juzen2003 committed:
- In menu.html, add bg-white class to list-group-item-action, this will
  remove the gray highlight when user mouses over to the blue text under
  surface geometry and ring geometry.

dstopp commented on opus/application/static_media/js/opus.js:
```diff
@@ -646,13 +659,17 @@ var opus = {
                 // ESC key - close modals and help panel
-                $(".op-confirm-modal").modal('hide');
+                // But don't close "#op-browser-version-msg" and "#op-browser-size-msg"
+                if (!$(".op-confirm-modal").is("#op-browser-version-msg") &&
```
i wonder if it would be useful in the future to have a class or data element on the modal that identifies it as "close-able on esc", instead of having to check the id of the modal.    Then you could just do something like:
if ($(".op-confirm-modal").data("action") === "esc")    more or less, doesn't have to be exactly this.

dstopp commented on opus/application/static_media/js/opus.js:
```diff
@@ -667,7 +684,7 @@ var opus = {
                             break;
                     }
-                    $(".modal").modal("hide");
+                    $(`#${target}`).modal("hide");
```
don't you need the same fix for the "cancel"...?

dstopp commented on opus/application/static_media/js/opus.js:
```diff
@@ -646,13 +659,17 @@ var opus = {
 
         // Handle the Submit or Cancel buttons for the various confirm modals we can pop up
-        $(".op-confirm-modal").on("click", ".btn", function() {
+        $(".op-confirm-modal").on("click", ".btn", function(event) {
```
while i think it's good form to add the parameter event, i'm not sure i see where you are using it...  I also think it might be good to standardize on 'e' vs 'event' as a name, we use both.

dstopp commented on opus/application/static_media/js/opus.js:
```diff
@@ -724,12 +741,112 @@ var opus = {
+            matchObj = userAgent.match(/Version\/(\d+.\d+)/);
+            browserName = "Safari";
+            browserVersion = matchObj[1];
+        }
```
what if it's none of the "else if" group....?

juzen2003 committed:
- In confirm_modal.html, add data-action to store "esc" (take in
  close_by_keyboard parameter) when confirmation modal is created. This
  data attribute will be used in opus.js keydown event handler (line 663)
  to determine if "esc" key will hide the modal.
- In base.html, pass in close_by_keyboard="esc" if the confirm modal can
  be closed by "esc" key.
- In opus.js isBrowserSupported function, set browserName to "unsupported"
  , and browserVersion to "0.0" if browser is not in the supported list.
- In opus.js, remove "event" in .op-confirm-modal button click event
  handler callback function because it's not used.
- In opus.js .op-confirm-modal button click event handler, hide the modal
  by $(`#${target}`).modal("hide") instead of $(".modal").modal("hide").
  This will make sure only the target modal is closed. (not all the modals
  with .modal class)

juzen2003 committed:
- In opus.js, update browserSupport with the latest supported browser
  version and window size.
- In opus.js, update isBrowserSupported, remove the if else block for
  edge (because it's not supported). Also update the string interpolation
  for modalMsg (line 789-798) so that we only need to change browserSupport
  in the future, and modalMsg will get updated automatically.
- In about.html, update the supported browser versions in the messages.

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -794,10 +786,16 @@ var opus = {
+            browser = browserName;
+        }
+        let modalMsg = (`Your current browser is ${browser}.
+                        Please update the browser. OPUS supports
```
One more change here. Right now it will say "Your current browser is unsupported. Please update the browser." That doesn't make much sense. Move the "Please update the browser." into the let statement at 789 and then it will work better.

Also let's change it to "Please update your browser."

rfrenchseti commented on opus/application/apps/help/templates/help/about.html:
```diff
@@ -177,7 +177,7 @@
 </p>
 
 <p>Please note that OPUS works best with a browser size of at least 1280 by 720. We
```
This message needs to be changed to the new sizes, doesn't it?

juzen2003 committed:
- In about.html, update the messages saying that "OPUS works best with a
  browser size of at least 600 by 200."
- In opus.js, inside isBrowserSupported, update the string interpolation
  for variables: browser and modalMsg.
